chainstate/test-suite/src/tests/pos_processing_tests.rs

```diff
@@ -0,0 +1,401 @@
+        );
+
+        let res = tf
+            .make_block_builder()
+            .with_consensus_data(ConsensusData::PoS(pos_data))
+            .build_and_process()
+            .unwrap_err();
+
+        assert_eq!(res, expected_error);
+    }
+
+    {
+        // invalid timestamp
+        let prev_randomness = tf.chainstate.get_chain_config().initial_randomness();
+        let block_timestamp =
+            BlockTimestamp::from_duration_since_epoch(tf.current_time().saturating_mul(2));
+        let vrf_transcript = construct_transcript(1, &prev_randomness, block_timestamp);
+        let vrf_data = vrf_sk.produce_vrf_data(vrf_transcript.into());
+        let pos_data = PoSData::new(
+            vec![TxInput::from_outpoint(stake_pool_outpoint.clone())],
+            vec![InputWitness::NoSignature(None)],
+            pool_id,
+            vrf_data,
+            Compact::from(difficulty),
+        );
+
+        let res = tf
+            .make_block_builder()
+            .with_consensus_data(ConsensusData::PoS(pos_data))
+            .build_and_process()
+            .unwrap_err();
+
+        assert_eq!(res, expected_error);
+    }
+
+    {
+        // invalid epoch
+        let prev_randomness = tf.chainstate.get_chain_config().initial_randomness();
+        let block_timestamp = BlockTimestamp::from_duration_since_epoch(tf.current_time());
+        let vrf_transcript = construct_transcript(2, &prev_randomness, block_timestamp);
+        let vrf_data = vrf_sk.produce_vrf_data(vrf_transcript.into());
+        let pos_data = PoSData::new(
+            vec![TxInput::from_outpoint(stake_pool_outpoint.clone())],
+            vec![InputWitness::NoSignature(None)],
+            pool_id,
+            vrf_data,
+            Compact::from(difficulty),
+        );
+
+        let res = tf
+            .make_block_builder()
+            .with_consensus_data(ConsensusData::PoS(pos_data))
+            .build_and_process()
+            .unwrap_err();
+
+        assert_eq!(res, expected_error);
+    }
+
+    {
+        // invalid vrf private key
+        let prev_randomness = tf.chainstate.get_chain_config().initial_randomness();
+        let block_timestamp = BlockTimestamp::from_duration_since_epoch(tf.current_time());
+        let vrf_transcript = construct_transcript(1, &prev_randomness, block_timestamp);
+        let (vrf_sk_2, _) = VRFPrivateKey::new(VRFKeyKind::Schnorrkel);
+        let vrf_data = vrf_sk_2.produce_vrf_data(vrf_transcript.into());
+        let pos_data = PoSData::new(
+            vec![TxInput::from_outpoint(stake_pool_outpoint)],
+            vec![InputWitness::NoSignature(None)],
+            pool_id,
+            vrf_data,
+            Compact::from(difficulty),
+        );
+
+        let res = tf
+            .make_block_builder()
+            .with_consensus_data(ConsensusData::PoS(pos_data))
+            .build_and_process()
+            .unwrap_err();
+
+        assert_eq!(res, expected_error);
+    }
+}
```

Comment:
Always when you test multiple invalid cases, end it with a valid case to ensure that your reference case works.